derive-impl/Cargo.toml

```diff
@@ -13,7 +13,7 @@ proc-macro = true
 [dependencies]
 proc-macro2 = { version = "1.0.24", features = ["span-locations"] }
 proc-macro-crate = "3.1.0"
-proc-macro-error = "1.0.4"
+proc-macro-error2 = "2.0.1"
```

Comment:
This crate doesn't seem too popular. What's the issue with keeping the old one?

Reply:
The reason is https://rustsec.org/advisories/RUSTSEC-2024-0370.html. And if you run something like cargo-deny (which we do on CI), it will complain about it. And I would expect that maybe someone will complain, hence it's easier to upgrade to that one. Another point is that crates are moving on to `syn` v2, so you would've less duplicated deps as `proc-macro-error` is still using v1.

Reply:
I'd be more concerned about using a random crate than using a so-called "unmaintained" crate, from a supply-chain perspective at least.

Reply:
The supply-chain perspective is a good point. I'll update that PR when I find time.

Reply:
(I have to wonder if we could just vendor it)

Reply:
Looks like a legit developer to me :shrug: 

Advisory warnings are annoying for downstream users for sure, so it is worth fixing one way or another.

Reply:
If you find it annoying, please complain to the rustsec project.

Reply:
To expand on that, it's one thing if an unmaintained project has security issues piling up. But jumping to the latest hot fork is just as likely to cause problems as not. Declaring a project unmaintained without a clear replacement is literally just dangling a vulnerable part of the OSS supply chain in front of potential attackers and asking them to "jump in and take a bite".

Reply:
> https://rustsec.org/advisories/RUSTSEC-2024-0370.html lists 3 replacements, including proc-macro-error2 used in this PR.

I'm talking about *clear* replacements. In this case, both the advisory and the suggestion to use `proc-macro-error2` were submitted by the author of `proc-macro-error2`.

The other alternatives were added later, but still aren't clear replacements.

> So overall I think it is fair to warn users that this crate is unmaintained and if they have any issues with it in the future, there will be no easy solution for them, which is even worse when dependency is indirect.

RUSTSEC is a security project and treats these advisories as security advisories. Sure, I get warning about unmaintained security critical bits _if_ there is a clear maintained replacement, but reports like this one just cause people to complain and are more likely to _cause_ security issues than they are to resolve them.

Reply:
Ok, here's a version that simply drops the offending dependency: https://github.com/multiformats/rust-multihash/pull/372.